crush/crush.go

```diff
@@ -101,6 +103,38 @@ func CreateTarGz(destination io.Writer, source string) error {
 	return CreateTar(gz, source)
 }
 
+// ExtractArchive extracts source archive to a destination directory.  An arbitrary number of top-level directory
+// components can be stripped from each path.
+func ExtractArchive(source *os.File, destination string, stripComponents int) error {
+	b, err := ioutil.ReadFile(source.Name())
+	if err != nil {
+		return err
+	}
+	kind, err := filetype.Match(b)
+	if err != nil {
+		return err
+	}
```

Comment:
As I read this, it's loading the whole file into memory and passing that to Match. I don't think we'd want to do that because you have some archives which can be up to 500M. It's probably better to use one of the other matchers, like [MatchReader](https://github.com/h2non/filetype/blob/master/match.go#L55-L64) which seems to only read 8K.